### PR TITLE
feat: Adding options to make capability fields customizable

### PR DESCRIPTION
**Pull request by devotional-phoenix-97**
## Pull Request type

Please check the type of change your PR introduces:

- [ ] Bugfix
- [x] Feature
- [ ] Code style update (formatting, renaming)
- [ ] Refactoring (no functional changes, no API changes)
- [ ] Build-related changes
- [ ] Documentation content changes
- [ ] Other (please describe):

## What is the current behavior?

- Currently capabilities are customisable, but extra fields cannot be configured
- for argocd bootstrap app extra fields are hardcoded in api

## What is the new behavior?

- Config can specify extra fields for capabilities
- Extra fields supplied in PaaS wil be validated by regexp as specified in config
- Extra fields not supplied in PaaS wil be set to a default value as specified in config
- for argocd bootstrap app, data can still be set in hardcoded api fields but also in Extra fields
- Extra fields (when set in PaaS, or with defaults) are also copied to appset list generator entry

## Does this introduce a breaking change?

- [ ] Yes
- [x] No

## Other information

Removing ArgoCD hardcoded fields is API breaking and will be bundled with other API breaking changes on a future release with new API version.

### CODE REVIEW

**Comment by devotional-phoenix-97** [comment]
Still working on e2e tests and documentation

File: manifests/config/cm_opr-paas.yaml [hunk 1]
```diff
@@ -34,6 +34,19 @@ data:
...
+            - argo-service-argocd-application-controller
+        extra_fields:
+          gitUrl:
+            required: true
+            validation: "https://github.com/[a-z0-9-/]+/[a-z0-9-/]+"
+          gitRevision:
+            default: "main"
+            validation: ".*"
+          gitPath:
+            defaullt: "."
```
Comment (portly-halicore-76) [inline manifests/config/cm_opr-paas.yaml]: This typo will be caught once we move the paasConfig to a CRD. I'm however curious why this passes the current tests?

File: manifests/crds/cpet.belastingdienst.nl_paas.yaml [hunk 2]
```diff
@@ -42,12 +42,17 @@ spec:
               capabilities:
                 additionalProperties:
                   properties:
+                    custom_fields:
+                      additionalProperties:
+                        type: string
+                      description: Custom fields to configure this specific Capability
+                      type: object
                     enabled:
                       description: Do we want an ArgoCD namespace, default false
```
Comment (portly-halicore-76) [inline manifests/crds/cpet.belastingdienst.nl_paas.yaml]: This is not part of the changes but caught my eyes. Can you add a suitable description here?

File: api/v1alpha1/paas_types.go [hunk 3]
```diff
@@ -257,17 +259,50 @@ type PaasCapability struct {
...
+		"git_url":      pc.GitUrl,
+		"git_revision": pc.GitRevision,
+		"git_path":     pc.GitPath,
+	}
+	for key, value := range pc.CustomFields {
+		fields[key] = value
+	}
+	for key, fieldConf := range fieldConfig {
+		if value, exists := fields[key]; exists {
+			validation := ".*"
```
Comment (portly-halicore-76) [inline api/v1alpha1/paas_types.go]: Can you add this to the paasTypes description as the default validation being applied.

File: api/v1alpha1/paas_types.go [hunk 4]
```diff
@@ -257,17 +259,50 @@ type PaasCapability struct {
...
+			if fieldConf.Validation != "" {
+				validation = fieldConf.Validation
+			}
+			if matched, err := regexp.Match(validation, []byte(value)); err != nil {
```
Comment (portly-halicore-76) [inline api/v1alpha1/paas_types.go]: Nice!

File: test/e2e/fixtures/paas_config.yml [hunk 5]
```diff
@@ -10,6 +10,17 @@ argopermissions:
...
+      git_url:
+        validation: '^ssh:\/\/git@scm\/[a-zA-Z0-9-.\/]*.git$'
+        required: true
+      git_revision:
+        default: main
+      git_path:
+        default: '.'
+        validation: '^[a-zA-Z0-9.\/]*$'
+      test_field:
+        default: test123
```
Comment (portly-halicore-76) [inline test/e2e/fixtures/paas_config.yml]: Can we assert this in the argocd test as well?

File: test/e2e/fixtures/paas_config.yml [hunk 6]
```diff
@@ -10,6 +10,17 @@ argopermissions:
 capabilities:
   argocd:
     applicationset: argoas
+    custom_fields:
+      git_url:
+        validation: '^ssh:\/\/git@scm\/[a-zA-Z0-9-.\/]*.git$'
+        required: true
```
Comment (portly-halicore-76) [inline test/e2e/fixtures/paas_config.yml]: From the logic, it looks to my like required also requires a default to set when a value is missing. Can we assert that in a test case as well. Or is my assumption incorrect.

File: test/e2e/capabilityargocd_test.go [hunk 7]
```diff
@@ -26,21 +26,24 @@ const (
...
 
 func TestCapabilityArgoCD(t *testing.T) {
 	paasSpec := api.PaasSpec{
-		Requestor: "paas-requestor",
+		Requestor: paasRequestor,
 		Quota:     quota.Quotas{},
 		Capabilities: api.PaasCapabilities{
 			"argocd": api.PaasCapability{
+				CustomFields: map[string]string{
+					"git_revision": paasArgoGitRevision,
```
Comment (portly-halicore-76) [inline test/e2e/capabilityargocd_test.go:41]: Cool! We can still use the old spec combined with the new spec. Great solution!

**Review by portly-halicore-76 — CHANGES_REQUESTED** [review]
Great changes, nice feature. W'ld love to see some docs updated describing these changes, what can the customer configure with this. Some minor comments added.

File: internal/config/config.go [hunk 8]
```diff
@@ -147,10 +147,18 @@ func (caps ConfigCapabilities) Verify() []string {
...
+	QuotaSettings      ConfigQuotaSettings          `yaml:"quotas"`
+	ExtraPermissions   ConfigCapPerm                `yaml:"extra_permissions"`
+	DefaultPermissions ConfigCapPerm                `yaml:"default_permissions"`
+	CustomFields       map[string]ConfigCustomField `yaml:"custom_fields"`
+}
+
+type ConfigCustomField struct {
+	// How to validate input with a regular expression
+	Validation string `yaml:"validation"`
+	Default    string `yaml:"default"`
```
Comment (portly-halicore-76) [inline internal/config/config.go]: Can you add the other descriptions as well.

File: manifests/config/cm_opr-paas.yaml (hunk 1 above)
Comment (devotional-phoenix-97) [reply manifests/config/cm_opr-paas.yaml]: This is not config as used in e2e tests, but config as we ship it by default (ends up in install.yaml.
I have now updated it with values as we use them in e2e tests (except for e2e specific values).

File: manifests/crds/cpet.belastingdienst.nl_paas.yaml (hunk 2 above)
Comment (devotional-phoenix-97) [reply manifests/crds/cpet.belastingdienst.nl_paas.yaml]: Yeah, I noticed some more leftovers from implementing configurable capabilities.
I have added a separate new PR for it:
https://github.com/belastingdienst/opr-paas/pull/179

File: api/v1alpha1/paas_types.go (hunk 3 above)
Comment (devotional-phoenix-97) [reply api/v1alpha1/paas_types.go]: Done. And added more docs to config. And added TODO for when moving to PaasConfig.

File: test/e2e/fixtures/paas_config.yml (hunk 5 above)
Comment (devotional-phoenix-97) [reply test/e2e/fixtures/paas_config.yml]: Done

File: test/e2e/fixtures/paas_config.yml (hunk 6 above)
Comment (devotional-phoenix-97) [reply test/e2e/fixtures/paas_config.yml]: I changed this a bit. Now:
- when Required is set, Default is not used, and Paas must have value set.
- when Required is not set, value can be set in Paas, but if not it reverts to Default, which is '' when not set in config.

File: internal/config/config.go (hunk 8 above)
Comment (devotional-phoenix-97) [reply internal/config/config.go]: Yes, added...

**Comment by portly-halicore-76** [comment]
As discussed, please pass a single err back to paas(ns) reconciler. There the err will set the errCondition accordingly.

File: api/v1alpha1/paas_types.go [hunk 9]
```diff
@@ -267,17 +268,52 @@ type PaasCapability struct {
...
+				issues = append(issues, fmt.Errorf("Invalid value %s (does not match %s)", value, fieldConf.Validation))
+			}
+		} else if fieldConf.Required {
+			issues = append(issues, fmt.Errorf("Value %s is required", value))
+		} else {
+			fields[key] = fieldConf.Default
+		}
+	}
+	if len(issues) > 0 {
+		return nil, issues
```
Comment (portly-halicore-76) [inline api/v1alpha1/paas_types.go]:
```suggestion
		return nil, errors.Join(err...)
```

File: api/v1alpha1/paas_types.go [hunk 10]
```diff
@@ -267,17 +268,52 @@ type PaasCapability struct {
...
 	// They must be encrypted with the public key corresponding to the private key deployed together with the Paas operator
 	SshSecrets map[string]string `json:"sshSecrets,omitempty"`
-	// You can enable extra permissions for the service accounts beloning to this capability
+	// You can enable extra permissions for the service accounts belonging to this capability
 	// Exact definitions is configured in Paas Configmap
 	// Note that we want to remove (some of) these permissions in future releases (like self-provisioner)
 	ExtraPermissions bool `json:"extra_permissions,omitempty"`
 }
 
+func (pc *PaasCapability) CapExtraFields(fieldConfig map[string]ConfigCustomField) (fields map[string]string, issues []error) {
```
Comment (portly-halicore-76) [inline api/v1alpha1/paas_types.go]:
```suggestion
func (pc *PaasCapability) CapExtraFields(fieldConfig map[string]ConfigCustomField) (fields map[string]string, err error) {
```

File: internal/controller/argo_app.go [hunk 11]
```diff
@@ -71,6 +73,13 @@ func (r *PaasNSReconciler) backendArgoApp(
 	namespace := paasns.NamespaceName()
 	argoConfig := paas.Spec.Capabilities["argocd"]
 	argoConfig.SetDefaults()
+	fields, errs := argoConfig.CapExtraFields(GetConfig().Capabilities["argocd"].CustomFields)
+	if len(errs) > 0 {
```
Comment (portly-halicore-76) [inline internal/controller/argo_app.go]:
```suggestion
	if err != nil {
	return  nil, err
	}
```

File: internal/controller/capabilities.go [hunk 12]
```diff
@@ -122,27 +123,36 @@ func splitToService(paasName string) (string, string) {
...
+		// This function is called from on other function within this exact if exists check
+		return fmt.Errorf("We should never end here")
+	} else if capConfig, exists := GetConfig().Capabilities[paasns.Name]; !exists {
+		if setErr := r.setErrorCondition(ctx, paasns, err); setErr != nil {
+			return fmt.Errorf("while setting condition for %w, another error occurred: %w", err, setErr)
+		}
+		// Not needed to keep reconciling
 		return nil
+	} else if fields, errs = cap.CapExtraFields(capConfig.CustomFields); errs != nil {
+		if setErr := r.setErrorCondition(ctx, paasns, errors.Join(errs...)); setErr != nil {
```
Comment (portly-halicore-76) [inline internal/controller/capabilities.go]:
```suggestion
		if err != nil {
		   return err
		}
```

**Review by portly-halicore-76 — APPROVED** [review]
Thanks!

**Pull request merged**